node/src/store/indexed_db_store.rs

```diff
@@ -209,6 +221,64 @@ impl IndexedDbStore {
 
         height <= head_height
     }
+
+    async fn mark_header_sampled(
+        &self,
+        height: u64,
+        accepted: bool,
+        cids: Vec<Cid>,
+    ) -> Result<u64> {
+        // quick check with contains_height, which uses cached head
+        if !self.contains_height(height) {
+            return Err(StoreError::NotFound);
+        }
+
+        let tx = self
+            .db
+            .transaction(&[SAMPLING_STORE_NAME], TransactionMode::ReadWrite)?;
+        let sampling_store = tx.store(SAMPLING_STORE_NAME)?;
+
+        let metadata = SamplingMetadata {
+            accepted,
+            cids_sampled: cids,
+        };
+        let metadata_jsvalue = to_value(&metadata)?;
+        let height_key = to_value(&height)?;
+
+        sampling_store
+            .put(&metadata_jsvalue, Some(&height_key))
+            .await?;
+
+        tx.commit().await?;
+
+        let previously_lowest_unsampled = *self.lowest_unsampled_height.borrow();
+        let new_highest_sampled_height =
+            get_next_unsampled_height_from_database(&self.db, previously_lowest_unsampled).await?;
+        self.lowest_unsampled_height
+            .replace(new_highest_sampled_height);
```

Comment:
are we sure that this is concurrent safe?

Reply:
As I understand rust-wasm, since there's an await point between the read and write, it's indeed a race condition. 
I've separated the logic into `update_unsampled_height`, which checks whether the unsampled height was updated before write. This should result in behaviour equivalent to atomic compare/exchange (since we're running single threaded).